flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java

```diff
@@ -149,52 +150,72 @@ protected String getInitialShardIterator() throws Exception {
 
 		// before infinitely looping, we set the initial nextShardItr appropriately
 
-		if (lastSequenceNum.equals(SentinelSequenceNumber.SENTINEL_LATEST_SEQUENCE_NUM.get())) {
+		if (isSentinelSequenceNumber(lastSequenceNum)) {
+			nextShardItr = getShardIteratorForSentinel(lastSequenceNum);
+		} else {
+			// we will be starting from an actual sequence number (due to restore from failure).
+			// if the last sequence number refers to an aggregated record, we need to clean up any dangling sub-records
```

Comment:
comments here about record aggregation should go into the `getShardIteratorForRealSequenceNumber` method